### PR TITLE
[feat] Add support for schedulers

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR add a better support for optimizer and schedulers. 

Fixes # (issue)

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/core/model.py [hunk 1]
```diff
@@ -323,3 +343,67 @@ def available_models(cls) -> List[str]:
...
+        if isinstance(self.trainer.limit_train_batches, int) and self.trainer.limit_train_batches != 0:
+            dataset_size = self.trainer.limit_train_batches
+        elif isinstance(self.trainer.limit_train_batches, float):
+            # limit_train_batches is a percentage of batches
+            dataset_size = len(self.train_dataloader())
+            dataset_size = int(dataset_size * self.trainer.limit_train_batches)
+        else:
+            dataset_size = len(self.train_dataloader())
+
+        num_devices = max(1, self.trainer.num_gpus, self.trainer.num_processes)
```
Comment (SeanNaren) [inline flash/core/model.py:365]: Maybe someone else can confirm, but I think `num_gpus` is automatically set to `num_processes` now, so this isn't needed, we should be able to do just `num_devices = num_processes`

**Review by SeanNaren — APPROVED** [review]
Nice, might want to get someone else who hasn't seen this code to also review :)

File: flash/core/model.py [hunk 2]
```diff
@@ -168,8 +181,14 @@ def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> A
             batch = torch.stack(batch)
         return self(batch)
 
-    def configure_optimizers(self) -> torch.optim.Optimizer:
-        return self.optimizer_cls(filter(lambda p: p.requires_grad, self.parameters()), lr=self.learning_rate)
+    def configure_optimizers(self):
```
Comment (edgarriba) [inline flash/core/model.py:184]:
```suggestion
    def configure_optimizers(self) -> Union[Optimizer, Tuple[Optimizer, _LRScheduler]]:
```

File: flash/core/model.py [hunk 3]
```diff
@@ -323,3 +342,67 @@ def available_models(cls) -> List[str]:
...
+
+        if self.trainer.max_steps and self.trainer.max_steps < max_estimated_steps:
+            return self.trainer.max_steps
+        return max_estimated_steps
+
+    def _compute_warmup(self, num_training_steps: int, num_warmup_steps: Union[int, float]) -> int:
+        if isinstance(num_warmup_steps, float):
+            # Convert float values to percentage of training steps to use as warmup
+            num_warmup_steps *= num_training_steps
+        return int(num_warmup_steps)
```
Comment (edgarriba) [inline flash/core/model.py:379]: not sure how critical is this number, but using round might be more accurate.

File: flash/core/model.py [hunk 4]
```diff
@@ -323,3 +342,67 @@ def available_models(cls) -> List[str]:
...
+            return scheduler
+        if isinstance(scheduler, str):
+            scheduler_fn = self.schedulers.get(self.scheduler)
+            if "num_warmup_steps" in self.scheduler_kwargs:
+                num_warmup_steps = self.scheduler_kwargs.get("num_warmup_steps")
+                if not isinstance(num_warmup_steps, float) or (num_warmup_steps > 1 or num_warmup_steps < 0):
+                    raise MisconfigurationException(
+                        "`num_warmup_steps` should be provided as float between 0 and 1 in `scheduler_kwargs`"
+                    )
+                num_training_steps = self.get_num_training_steps()
```
Comment (edgarriba) [inline flash/core/model.py:393]:
```suggestion
                num_training_steps: int = self.get_num_training_steps()
```

File: flash/core/model.py [hunk 5]
```diff
@@ -323,3 +342,67 @@ def available_models(cls) -> List[str]:
...
+                num_warmup_steps = self._compute_warmup(
```
Comment (edgarriba) [inline flash/core/model.py:394]:
```suggestion
                num_warmup_steps: int = self._compute_warmup(
```

File: flash/core/model.py [hunk 6]
```diff
@@ -64,11 +68,16 @@ class Task(LightningModule):
         postprocess: :class:`~flash.data.process.Postprocess` to use as the default for this task.
     """
 
+    schedulers = _SCHEDULER_REGISTRY
```
Comment (edgarriba) [inline flash/core/model.py:71]:
```suggestion
    schedulers: FlashRegistry = _SCHEDULER_REGISTRY
```

File: flash/core/schedulers.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+from flash.core.registry import FlashRegistry
+from flash.utils.imports import _TRANSFORMERS_AVAILABLE
+
+_SCHEDULER_REGISTRY = FlashRegistry("scheduler")
+
+if _TRANSFORMERS_AVAILABLE:
+    from transformers import optimization
+    functions = [getattr(optimization, n) for n in dir(optimization) if ("get_" in n and n != 'get_scheduler')]
```
Comment (edgarriba) [inline flash/core/schedulers.py:8]:
```suggestion
    functions: List[Callable] = [getattr(optimization, n) for n in dir(optimization) if ("get_" in n and n != 'get_scheduler')]
```

File: flash/core/schedulers.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
+from flash.core.registry import FlashRegistry
+from flash.utils.imports import _TRANSFORMERS_AVAILABLE
+
+_SCHEDULER_REGISTRY = FlashRegistry("scheduler")
+
+if _TRANSFORMERS_AVAILABLE:
+    from transformers import optimization
```
Comment (ethanwharris) [inline flash/core/schedulers.py:7]: Why only add transformers optimizers here? Maybe also add `torch.optim` optimizers?

File: flash/utils/imports.py [hunk 9]
```diff
@@ -5,3 +5,8 @@
 _COCO_AVAILABLE = _module_available("pycocotools")
 _TIMM_AVAILABLE = _module_available("timm")
 _TORCHVISION_AVAILABLE = _module_available("torchvision")
+try:
+    import transformers
+    _TRANSFORMERS_AVAILABLE = True
```
Comment (ethanwharris) [inline flash/utils/imports.py:10]: Isn't this what `_module_available` does?

**Review by ethanwharris — COMMENTED** [review]
Overall, looks good :smiley: Can you maybe add `if _TRANSFORMERS_AVAILABLE` to the other files that use transformers (e.g. `/text/seq2seq/summarization/data.py`)?

File: flash/core/model.py (hunk 2 above)
Comment (tchaton) [reply flash/core/model.py:184]: Done !

File: flash/core/model.py (hunk 3 above)
Comment (tchaton) [reply flash/core/model.py:379]: Yes, definitely.

File: flash/core/model.py (hunk 5 above)
Comment (tchaton) [reply flash/core/model.py:394]: Done !

File: flash/core/model.py (hunk 6 above)
Comment (tchaton) [reply flash/core/model.py:71]: Done !

File: flash/core/schedulers.py (hunk 7 above)
Comment (tchaton) [reply flash/core/schedulers.py:8]: Done !

File: flash/core/schedulers.py (hunk 8 above)
Comment (tchaton) [reply flash/core/schedulers.py:7]: transformers already provide lambda lr. Users can provide their own scheduler if needed. Registry should be more custom things IMO.

File: flash/utils/imports.py (hunk 9 above)
Comment (tchaton) [reply flash/utils/imports.py:10]: Somehow, it doesn't work with transformers as spec isn't available.

File: flash/core/schedulers.py (hunk 8 above)
Comment (ethanwharris) [reply flash/core/schedulers.py:9]: Ok, I'm a little confused. The user can just provide a scheduler from transformers too right?

File: flash/utils/imports.py (hunk 9 above)
Comment (ethanwharris) [reply flash/utils/imports.py]: That's strange, `_module_available("transformers")` is giving true for me but I haven't tried it without transformers installed

**Review by ethanwharris — APPROVED** [review]
